devlib/trace/systrace.py

```diff
@@ -0,0 +1,137 @@
+    :type outdir: str
+
+    :param categories: Systrace categories to trace. See `available_categories`
+    :type categories: list(str)
+
+    :param buffer_size: Buffer size in kb
+    :type buffer_size: int
+
+    :param strict: Raise an exception if any of the requested categories
+        are not available
+    :type strict: bool
+    """
+
+    @property
+    @memoized
+    def available_categories(self):
+        lines = subprocess.check_output([self.systrace_binary, '-l']).splitlines()
+
+        categories = []
+        for line in lines:
+            categories.append(line.split()[0])
+
+        return categories
+
+    def _build_cmd(self):
+        self.systrace_cmd = '{} -o {} -e {}'.format(
+            self.systrace_binary,
+            self._outfile,
+            self.target.adb_name
+        )
+
+        if self.buffer_size:
+            self.systrace_cmd += ' -b {}'.format(self.buffer_size)
+
+        self.systrace_cmd += ' {}'.format(' '.join(self.categories))
+
+    def __init__(self, target,
+                 outdir,
+                 categories=None,
+                 buffer_size=None,
+                 strict=False):
+
+        super(SystraceCollector, self).__init__(target)
+
+        if not os.path.isdir(outdir):
+            raise RuntimeError('"outdir" parameter is not a valid directory')
+
+        self._outfile = os.path.join(outdir, 'trace.html')
+        self.categories = categories or DEFAULT_CATEGORIES
+        self.buffer_size = buffer_size
+
+        # Try to find a systrace binary
+        self.systrace_binary = None
+
+        systrace_binary_path = os.path.join(platform_tools, 'systrace', 'systrace.py')
+        if not os.path.isfile(systrace_binary_path):
+            raise HostError('Could not find any systrace binary under {}'.format(platform_tools))
+
+        self.systrace_binary = systrace_binary_path
+
+        # Filter the requested categories
+        for category in self.categories:
+            if category not in self.available_categories:
+                message = 'Category [{}] not available for tracing'.format(category)
+                if strict:
+                    raise TargetError(message)
+                self.logger.warning(message)
+
+        self.categories = list(set(self.categories) & set(self.available_categories))
+        if not self.categories:
+            raise TargetError('None of the requested categories are available')
+
+        self._build_cmd()
+
+    def reset(self):
+        self.stop()
+
+    def start(self):
+        self._systrace_process = subprocess.Popen(
+            self.systrace_cmd,
+            stdin=subprocess.PIPE,
+            shell=True
+        )
+
+    def stop(self):
+        # Systrace expects <enter> to stop
+        if self._systrace_process:
+            self._systrace_process.communicate('\n')
+            self._systrace_process = None
+
+    def get_trace(self, outfile):
+        copyfile(self._outfile, outfile)
```

Comment:
Don't we want a copy ? For instance with Ftrace we can call get_trace() as many times as we want since we pull the trace file from the target. Seems like `SerialTraceCollector` behaves like this too (and uses a tempfile... I should've looked at this one earlier :D)

Reply:
Wait no actually SerialTraceCollector is a one-time copy. I suppose either is fine then ?

Reply:
A move is preferable if the files are expected to be fairly large, but other than that, either is fine.